cai_causal_graph/causal_graph.py

```diff
@@ -534,6 +542,75 @@ def is_empty(self) -> bool:
         """Return True if there are no nodes and edges. False otherwise."""
         return len(self.nodes) == 0 and len(self.edges) == 0
 
+    def _check_node(self, identifier: NodeLike) -> str:
+        identifier = Node.identifier_from(identifier)
+
+        if identifier in self._nodes_by_identifier:
+            raise CausalGraphErrors.NodeDuplicatedError(f'Node already exists: {identifier}')
+        return identifier
+
+    def _prepare_nodes(self, source: NodeLike, destination: NodeLike) -> Tuple[List[Node], List[Node]]:
+        """Prepare the source and destination nodes for adding an edge."""
+        source_meta = None
+        if isinstance(source, HasMetadata):
+            source_meta = source.get_metadata()
+
+        destination_meta = None
+        if isinstance(destination, HasMetadata):
+            destination_meta = destination.get_metadata()
+
+        source = Node.identifier_from(source)
+        destination = Node.identifier_from(destination)
+
+        # check that the source is not equal to destination
+        if source == destination:
+            raise CausalGraphErrors.CyclicConnectionError(
+                f'Adding an edge from {source} to {destination} would create a self-loop, no matter the edge type. '
+                f'This is currently not supported.'
+            )
+
+        source_nodes = self.get_nodes(source)
+        destination_nodes = self.get_nodes(destination)
```

Comment:
Is there a chance that there will ever be more than one source or destination? Curious why you're returning a list. Can TimeSeriesCausalGraphs return multiple nodes for the same identifier or something?